### PR TITLE
Backport PR #32166 on branch 1.0.x

### PR DESCRIPTION
**Pull request by jbrockmendel**
#32166

### CODE REVIEW

File: pandas/tests/indexes/timedeltas/test_join.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        )
+        str(df)
+
+        cols = df.columns.join(df.index, how="outer")
+        joined = cols.join(df.columns)
+        assert cols.dtype == np.dtype("O")
+        assert cols.dtype == joined.dtype
+        tm.assert_index_equal(cols, joined)
+
+    def test_join_preserves_freq(self):
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/timedeltas/test_join.py]: only this test should be backported

File: pandas/tests/indexes/datetimes/test_join.py [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+        msg = "tz-naive.*tz-aware"
+        with pytest.raises(TypeError, match=msg):
+            naive.join(aware)
+
+        with pytest.raises(TypeError, match=msg):
+            aware.join(naive)
+
+    @pytest.mark.parametrize("tz", [None, "US/Pacific"])
+    def test_join_preserves_freq(self, tz):
```
Comment (simonjayhawkins) [inline pandas/tests/indexes/datetimes/test_join.py]: only this test should be backported

File: pandas/core/indexes/datetimelike.py [hunk 3]
```diff
@@ -857,21 +857,16 @@ def _is_convertible_to_index_for_join(cls, other: Index) -> bool:
...
-            joined.name = name
-            return joined
-        else:
-            kwargs = {}
-            if hasattr(self, "tz"):
-                kwargs["tz"] = getattr(other, "tz", None)
-            return self._simple_new(joined, name, **kwargs)
+
+        freq = self.freq if self._can_fast_union(other) else None
+        new_data = type(self._data)._simple_new(  # type: ignore
```
Comment (simonjayhawkins) [inline pandas/core/indexes/datetimelike.py]: can you add a comment why the type:ignore has been added.

**Review by simonjayhawkins — CHANGES_REQUESTED** [review]
Thanks @jbrockmendel. Should the backport mirror the original PR more closely?

**Comment by jreback** [comment]
this is fine, has some extra tests which are nbd.

**Pull request merged**